kit/inputs/src/UIMaterialTextView/UIMaterialTextView.tsx

```
@@ -62,19 +59,36 @@ function useExtendedProps(
         isFocused,
     );
 
+    const { selectionEnd, onSelectionChange, skipNextOnSelectionChange } = useOnSelectionChange();
+
+    const applyMask = useApplyMask(mask, selectionEnd, skipNextOnSelectionChange);
+
     const { imperativeChangeText, moveCarret } = useImperativeChange(
         ref,
-        multiline,
-        onChangeTextWithInputHasValue,
+        onChangeTextProp,
+        checkInputHasValue,
+        applyMask,
     );
 
-    const { onChangeText, onSelectionChange } = useApplyMask(
-        imperativeChangeText,
-        moveCarret,
-        mask,
+    const onChangeText = React.useCallback(
+        (text: string) =>
+            imperativeChangeText(text, {
+                shouldSetNativeProps: false,
+            }),
+        [imperativeChangeText],
     );
 
-    const clear = useClear(resetInputHeight, onChangeText, ref);
+    React.useEffect(() => {
+        if (defaultValue) {
+            imperativeChangeText(defaultValue);
```

Hm, why it's done with useEffect? 🤔 isn't it better to do on the first render (with ref as a guard) and put it as a `defaultValue` to nested `TextInput`?

---

We need to call imperativeChangeText on first the first render to validate defaultValue with mask and to calculate other fields in this component (e.g. `checkInputHasValue` and dependent on it).
Using useEffect with empty dependency array - is the simpliest way to do this on first render.

---

But how do you guarantee a consistency between mount and `useEffect`?

---

Because between those two states input will have an incorrect value

---

Why I need guarantee this consistency? Update takes tiny time. I don't think we need to make it less. But this way guarantee ref to be assigned, that needed to proper work of this component.

---

1. useEffect doesn't guarantee that ref will be assigned.
2. on mount you have an incorrect `defaultValue`, that you try to override, it might blink to user, that leads to bad UI
3. how do you guarantee that it takes "tiny time"? I mean, that you tested it on your powerful machine doesn't make it very fast

---

I can use useLayoutEffect to make this calculations before first render without blink.
Putting defaultValue to nested TextInput is wrong because first of all defaultValue should be validated by mask.

---

`useLayoutEffect` doesn't make calculations before first render, in fact it actually can fire after `useEffect` :)

---

Done